### PR TITLE
Add support for using git submodules as subprojects

### LINKED ISSUE #1449: Auto-initialize submodules in subprojects/ directory as if it had a wrap file
The auto-init will only happen when a `subproject()` or `dependency(... fallback:)` call is encountered, just like a .wrap project.

Also we need to add an FAQ entry on how wraps actually work. A lot of people misunderstand them and think they will always fetch from the network.

Filing a bug about all this so I don't forget.

### PR DESCRIPTION
**Pull request by nirbheek**
Closes #1449.

Also adds a new option `--wrap-mode` that implements special wrap modes for use by packagers.

  nofallback: Don't download wraps for dependency() fallbacks
  nodownload: Don't download wraps for all subproject() calls

Subprojects are used for two purposes:
1. To download and build dependencies by using `.wrap` files if they are not provided by the system. This is usually expressed via `dependency(..., fallback: ...)`.
2. To download and build 'copylibs' which are meant to be used by copying into your project. This is always done with an explicit `subproject()` call.

`--wrap-mode=nofallback` will never do (1)
`--wrap-mode=nodownload` will do neither (1) nor (2)

If you are building from a release tarball, you should be able to safely use `nodownload` since upstream is expected to ship all required sources with the tarball.

If you are building from a git repository, you will want to use `nofallback` so that any 'copylib' wraps will be download as subprojects.

Note that these options do not affect subprojects that are git submodules since those are only usable in git repositories, and you almost always want to download them.

### CODE REVIEW

File: mesonbuild/mesonmain.py [hunk 1]
```diff
@@ -67,6 +68,10 @@ def add_builtin_argument(name, **kwargs):
                     help='Set project options.')
 parser.add_argument('-v', '--version', action='version',
                     version=coredata.version)
+                    # See the mesonlib.WrapMode enum for documentation
```
Comment (ignatenkobrain) [inline mesonbuild/mesonmain.py]: __[flake8]__
* [E116] unexpected indentation (comment)

File: mesonbuild/wrap/wrap.py [hunk 2]
```diff
@@ -86,29 +95,45 @@ def has_patch(self):
...
                 return packagename
-            else:
-                mlog.warning('Subproject directory %s is empty, possibly because of an unfinished'
-                             'checkout, removing to reclone' % dirname)
-                os.rmdir(dirname)
-        except NotADirectoryError:
-            raise RuntimeError('%s is not a directory, can not use as subproject.' % dirname)
-        except FileNotFoundError:
-            pass
+             # Is the dir not empty and also not a git submodule dir that is
```
Comment (ignatenkobrain) [inline mesonbuild/wrap/wrap.py]: __[flake8]__
* [E114] indentation is not a multiple of four (comment)

File: mesonbuild/wrap/wrap.py [hunk 3]
```diff
@@ -86,29 +95,45 @@ def has_patch(self):
...
+             # not checkout properly? Can't do anything, exception!
```
Comment (ignatenkobrain) [inline mesonbuild/wrap/wrap.py]: __[flake8]__
* [E114] indentation is not a multiple of four (comment)

**Comment by nirbheek** [comment]
The `--wrap-mode` has come up repeatedly in discussions with various distro packagers (Fedora, Gentoo, Debian, etc), and has even come up on the [Mesa-dev mailing list](https://lists.freedesktop.org/archives/mesa-dev/2017-March/149050.html):

> I have no plans to implement wrap for mesa in the port I would do, and would NAK any patches that used wrap on Linux or BSD

This is an understandable sentiment, but there is no need to enforce this with policy. This option removes the need to have such a policy.

A lot of people seem to misunderstand the purpose of wraps and the control Meson has over them. The implementation of this option is both needed and will demonstrate that yes, we do actually know and care that a lot of people *need* the ability to completely disable automagic network downloads.

**Pull request merged**